### PR TITLE
Start adding OpenMM support

### PR DESCRIPTION
**Pull request by drroe**
This MR adds the ability to link in OpenMM to cpptraj. It also adds some limited ability to do minimization either using OpenMM to calculate forces or some built-in routines (unconstrained bonds, angles, dihedrals w/ 1-4 terms, and a very basic nonbonded term). The angle and dihedral terms were adapted from AmberTools SFF. Right now these terms and the minimizer will match SANDER from Amber exactly (given the initial structure is read in double precision, i.e. as a reference COORDS set). 

Currently this is only relevant for the new `emin` action, which is still under development. However, I have been using it in my research (it's handy for doing small relaxations), and I plan on making much more heavy use of the OpenMM backend soon, so I'm going to get this committed.

### CODE REVIEW

**Comment by swails** [comment]
If you merge my pull request, you can get Travis replaced with GitHub Actions. Builds finish in 15 minutes on that platform so it's much better.

File: src/MdOpts.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  }
+  scaleEE_ = argIn.getKeyDouble("scaleee", 1.0/1.2);
+  scaleNB_ = argIn.getKeyDouble("scalenb", 1.0/2.0);
+  shakeType_ = Constraints::OFF;
+  std::string shakearg = argIn.GetStringKey("shake");
+  if (!shakearg.empty()) {
+    if (shakearg == "h" || shakearg == "hydrogen")
+      shakeType_ = Constraints::BONDS_TO_H;
+    else if (shakearg == "all")
+      shakeType_ = Constraints::ALL_BONDS;
```
Comment (swails) [inline src/MdOpts.cpp:49]: I think openmm also supports h angles if that's worth supporting here.

File: src/MdOpts.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  public:
+    MdOpts();
+
+    int GetOptsFromArgs(ArgList&);
+
+    void PrintOpts() const;
+
+    Constraints::ShakeType Shake() const { return shakeType_; }
+    double ScaleEE()       const { return scaleEE_; }
+    double ScaleNB()       const { return scaleNB_; }
```
Comment (swails) [inline src/MdOpts.h:14]: Do you support mixed scaling like for glycam-amber combos?

File: src/PotentialTerm_OpenMM.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+  // Add angles
+  AddAngles(angleStretch, topIn.Angles(), topIn.AngleParm(), oldToNew);
+  AddAngles(angleStretch, topIn.AnglesH(), topIn.AngleParm(), oldToNew);
+
+  // Add dihedrals
+  AddDihedrals(ptorsion, topIn.Dihedrals(), topIn.DihedralParm(), oldToNew);
+  AddDihedrals(ptorsion, topIn.DihedralsH(), topIn.DihedralParm(), oldToNew);
+
+  // Populate nonbonded exclusions
+  nonbond->createExceptionsFromBonds(bondPairs, scaleEE_, scaleNB_);
```
Comment (swails) [inline src/PotentialTerm_OpenMM.cpp:99]: After this command you can go back and add exceptions for each 1-4 term individually in order to implement mixed 1-4 scaling. 

ParmEd has logic to do this in structure.py if you want to see a reference that has been validated. 

Just make sure you mark the exception to replace an existing one or you'll get an error.

**Review by swails — COMMENTED** [review]
Very cool! What do you think of OpenMM??

File: src/PotentialTerm_OpenMM.cpp (hunk 3 above)
Comment (swails) [reply src/PotentialTerm_OpenMM.cpp:99]: I actually wrote a C++-based Amber topology file parser + OpenMM system generator before I left for Lutron, and I transferred it to Todd Martinez so he could use it for something.

You can see the implementation here that obeys mixed 1-4 scaling in Amber topology files: https://github.com/tmartine68/omm_amber_cpp/blob/master/src/amberparm.cpp#L564-L584

**Comment by swails** [comment]
> If you merge my pull request, you can get Travis replaced with GitHub Actions. Builds finish in 15 minutes on that platform so it's much better.

I can also help pull in OpenMM in one of the builds to test the new functionality (as well as pulling it into the Jenkins recipe).  Added incentive to merge #864 :)

File: src/MdOpts.cpp (hunk 1 above)
Comment (drroe) [reply src/MdOpts.cpp:49]: Cool - eventually I would like to add support for all features supported by OpenMM - just wanted to get the initial stuff in the master branch so things don't diverge too much.

File: src/MdOpts.h (hunk 2 above)
Comment (drroe) [reply src/MdOpts.h:14]: No, right now OpenMM support is very basic, so only the universal scaling is supported :-(

File: src/PotentialTerm_OpenMM.cpp (hunk 3 above)
Comment (drroe) [reply src/PotentialTerm_OpenMM.cpp:99]: Nice, I didn't know about the exception stuff! I guess I should have tapped you to write the OpenMM support for CPPTRAJ! 😊

**Comment by drroe** [comment]
Yay - appveyor finally fixed!

**Comment by drroe** [comment]
Uh oh:
```
Cannot contact Wonder Woman-cuda: hudson.remoting.ChannelClosedException: Channel "hudson.remoting.Channel@140592d6:Wonder Woman-cuda": Remote call on Wonder Woman-cuda failed. The channel is closing down or has closed down
```
I'll try restarting.

**Pull request merged**